hera_qm/metrics_io.py

```diff
@@ -0,0 +1,640 @@
+
+    Adds allowed types to the current group as datasets
+    creates new subgroups if a given item in a dictionary is a dictionary.
+
+    Will add attributes for HDF5 objects as follows
+        Datasets:
+            key_is_string: Boolean flag to determine if dictionary key is a string
+
+        Groups:
+            group_is_ordered: Boolean flag to determine if dict was an OrderedDict
+
+    Arguments
+        h5file: H5py file object into which data is written, this is edited in place.
+        path: absolute path in HDF5 to store the current dictionary
+        in_dict: Dictionary to be recursively walked and stored in h5file
+    Returns
+        None
+    """
+    allowed_types = (np.ndarray, np.float, np.int, bytes, six.text_type, list)
+    compressable_types = (np.ndarray, list)
+    for key in in_dict:
+        key_str = str(key)
+        if key == 'reds':
+            in_dict[key] = _reds_list_to_dict(in_dict[key])
+
+        if isinstance(in_dict[key], allowed_types):
+
+            if key in antpol_keys:
+                in_dict[key] = np.array(in_dict[key],
+                                        dtype=antpol_dtype)
+            if key in antpair_keys:
+                in_dict[key] = np.array(in_dict[key],
+                                        dtype=antpair_dtype)
+
+            if isinstance(in_dict[key], compressable_types):
+                try:
+                    dset = h5file[path].create_dataset(key_str,
+                                                       data=in_dict[key],
+                                                       compression='lzf')
+                    # Add boolean attribute to determine if key is a string
+                    # Used to parse keys saved to dicts when reading
+                    dset.attrs['key_is_string'] = isinstance(key, str)
+                except TypeError as err:
+                    raise TypeError("Input dictionary key: {0} does not have "
+                                    "compatible dtype. Received this error: "
+                                    "{1}".format(key, err))
+            else:
+                dset = h5file[path].create_dataset(key_str, data=in_dict[key])
+                # Add boolean attribute to determine if key is a string
+                # Used to parse keys saved to dicts when reading
+                dset.attrs['key_is_string'] = isinstance(key, str)
+
+        elif isinstance(in_dict[key], dict):
+            grp = h5file[path].create_group(key_str)
+            # Add boolean attribute to determine if input dictionary
+            # was and OrderedDict Orde
+            grp.attrs['group_is_ordered'] = False
+            if isinstance(in_dict[key], OrderedDict):
+                grp.attrs['group_is_ordered'] = True
+
+                # Generate additional dataset in an ordered group to save
+                # the order of the group
+                key_order = list(map(str, list(in_dict[key].keys())))
```

Comment:
in_dict[keys].keys() is a list in python 2 and a generator in python 3. I actually am not sure if it is needed or not though. If map() works on a generator then it could be dropped. I'll look into it.